File: java-module/ui-selenium/src/main/java/com/enterprise/qa/ui/pages/BasePage.java
package com.enterprise.qa.ui.pages;

import com.enterprise.qa.core.ai.selfhealing.SelfHealingLocator;
import com.enterprise.qa.core.config.ConfigManager;
import com.enterprise.qa.core.drivers.DriverManager;
import com.enterprise.qa.ui.helpers.ElementHelper;
import com.enterprise.qa.ui.helpers.SyncHelper;
import io.qameta.allure.Step;
import lombok.extern.slf4j.Slf4j;
import org.openqa.selenium.*;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.support.PageFactory;
import org.openqa.selenium.support.ui.Select;

import java.util.List;

/**
 * Base page object class providing common functionality for all page objects.
 * Includes self-healing locator support and comprehensive element interaction methods.
 */
@Slf4j
public abstract class BasePage {

    protected final WebDriver driver;
    protected final SelfHealingLocator selfHealingLocator;
    protected final ElementHelper elementHelper;
    protected final SyncHelper syncHelper;
    protected final ConfigManager config;

    public BasePage() {
        this.driver = DriverManager.getDriver();
        this.selfHealingLocator = new SelfHealingLocator(driver);
        this.elementHelper = new ElementHelper(driver);
        this.syncHelper = new SyncHelper(driver);
        this.config = ConfigManager.getInstance();
        PageFactory.initElements(driver, this);
    }

    /**
     * Gets the page title.
     *
     * @return the current page title
     */
    public String getPageTitle() {
        return driver.getTitle();
    }

    /**
     * Gets the current URL.
     *
     * @return the current URL
     */
    public String getCurrentUrl() {
        return driver.getCurrentUrl();
    }

    /**
     * Navigates to a URL.
     *
     * @param url the URL to navigate to
     */
    @Step("Navigate to: {url}")
    public void navigateTo(String url) {
        log.info("Navigating to: {}", url);
        driver.get(url);
        waitForPageLoad();
    }

    /**
     * Refreshes the current page.
     */
    @Step("Refresh page")
    public void refreshPage() {
        log.debug("Refreshing page");
        driver.navigate().refresh();
        waitForPageLoad();
    }

    /**
     * Navigates back to the previous page.
     */
    @Step("Navigate back")
    public void navigateBack() {
        log.debug("Navigating back");
        driver.navigate().back();
    }

    /**
     * Navigates forward to the next page.
     */
    @Step("Navigate forward")
    public void navigateForward() {
        log.debug("Navigating forward");
        driver.navigate().forward();
    }

    /**
     * Finds an element with self-healing capability.
     *
     * @param locator     the element locator
     * @param elementName a descriptive name for the element
     * @return the found element
     */
    protected WebElement findElement(By locator, String elementName) {
        return selfHealingLocator.findElement(locator, elementName);
    }

    /**
     * Finds multiple elements with self-healing capability.
     *
     * @param locator     the element locator
     * @param elementName a descriptive name for the elements
     * @return list of found elements
     */
    protected List<WebElement> findElements(By locator, String elementName) {
        return selfHealingLocator.findElements(locator, elementName);
    }

    /**
     * Clicks on an element.
     *
     * @param element     the element to click
     * @param elementName a descriptive name for the element
     */
    @Step("Click on: {elementName}")
    protected void click(WebElement element, String elementName) {
        syncHelper.waitForElementClickable(element);
        log.debug("Clicking on: {}", elementName);
        element.click();
    }

    /**
     * Clicks on an element by locator.
     *
     * @param locator     the element locator
     * @param elementName a descriptive name for the element
     */
    @Step("Click on: {elementName}")
    protected void click(By locator, String elementName) {
        WebElement element = findElement(locator, elementName);
        click(element, elementName);
    }

    /**
     * Double-clicks on an element.
     *
     * @param element     the element to double-click
     * @param elementName a descriptive name for the element
     */
    @Step("Double-click on: {elementName}")
    protected void doubleClick(WebElement element, String elementName) {
        syncHelper.waitForElementClickable(element);
        log.debug("Double-clicking on: {}", elementName);
        new Actions(driver).doubleClick(element).perform();
    }

    /**
     * Right-clicks (context clicks) on an element.
     *
     * @param element     the element to right-click
     * @param elementName a descriptive name for the element
     */
    @Step("Right-click on: {elementName}")
    protected void rightClick(WebElement element, String elementName) {
        syncHelper.waitForElementClickable(element);
        log.debug("Right-clicking on: {}", elementName);
        new Actions(driver).contextClick(element).perform();
    }

    /**
     * Types text into an element.
     *
     * @param element     the element to type into
     * @param text        the text to type
     * @param elementName a descriptive name for the element
     */
    @Step("Type '{text}' into: {elementName}")
    protected void type(WebElement element, String text, String elementName) {
        syncHelper.waitForElementVisible(element);
        log.debug("Typing '{}' into: {}", text, elementName);
        element.clear();
        element.sendKeys(text);
    }

    /**
     * Types text into an element by locator.
     *
     * @param locator     the element locator
     * @param text        the text to type
     * @param elementName a descriptive name for the element
     */
    @Step("Type '{text}' into: {elementName}")
    protected void type(By locator, String text, String elementName) {
        WebElement element = findElement(locator, elementName);
        type(element, text, elementName);
    }

    /**
     * Clears an input field.
     *
     * @param element     the element to clear
     * @param elementName a descriptive name for the element
     */
    @Step("Clear: {elementName}")
    protected void clear(WebElement element, String elementName) {
        syncHelper.waitForElementVisible(element);
        log.debug("Clearing: {}", elementName);
        element.clear();
    }

    /**
     * Gets the text of an element.
     *
     * @param element     the element
     * @param elementName a descriptive name for the element
     * @return the element's text
     */
    protected String getText(WebElement element, String elementName) {
        syncHelper.waitForElementVisible(element);
        String text = element.getText();
        log.debug("Got text '{}' from: {}", text, elementName);
        return text;
    }

    /**
     * Gets an attribute value of an element.
     *
     * @param element     the element
     * @param attribute   the attribute name
     * @param elementName a descriptive name for the element
     * @return the attribute value
     */
    protected String getAttribute(WebElement element, String attribute, String elementName) {
        syncHelper.waitForElementVisible(element);
        return element.getAttribute(attribute);
    }

    /**
     * Checks if an element is displayed.
     *
     * @param element the element to check
     * @return true if displayed
     */
    protected boolean isDisplayed(WebElement element) {
        try {
            return element.isDisplayed();
        } catch (NoSuchElementException | StaleElementReferenceException e) {
            return false;
        }
    }

    /**
     * Checks if an element is enabled.
     *
     * @param element the element to check
     * @return true if enabled
     */
    protected boolean isEnabled(WebElement element) {
        try {
            return element.isEnabled();
        } catch (NoSuchElementException | StaleElementReferenceException e) {
            return false;
        }
    }

    /**
     * Checks if a checkbox or radio button is selected.
     *
     * @param element the element to check
     * @return true if selected
     */
    protected boolean isSelected(WebElement element) {
        try {
            return element.isSelected();
        } catch (NoSuchElementException | StaleElementReferenceException e) {
            return false;
        }
    }

    /**
     * Selects an option from a dropdown by visible text.
     *
     * @param element     the select element
     * @param text        the option text
     * @param elementName a descriptive name for the element
     */
    @Step("Select '{text}' from: {elementName}")
    protected void selectByVisibleText(WebElement element, String text, String elementName) {
        syncHelper.waitForElementVisible(element);
        log.debug("Selecting '{}' from: {}", text, elementName);
        new Select(element).selectByVisibleText(text);
    }

    /**
     * Selects an option from a dropdown by value.
     *
     * @param element     the select element
     * @param value       the option value
     * @param elementName a descriptive name for the element
     */
    @Step("Select value '{value}' from: {elementName}")
    protected void selectByValue(WebElement element, String value, String elementName) {
        syncHelper.waitForElementVisible(element);
        log.debug("Selecting value '{}' from: {}", value, elementName);
        new Select(element).selectByValue(value);
    }

    /**
     * Selects an option from a dropdown by index.
     *
     * @param element     the select element
     * @param index       the option index
     * @param elementName a descriptive name for the element
     */
    @Step("Select index {index} from: {elementName}")
    protected void selectByIndex(WebElement element, int index, String elementName) {
        syncHelper.waitForElementVisible(element);
        log.debug("Selecting index {} from: {}", index, elementName);
        new Select(element).selectByIndex(index);
    }

    /**
     * Hovers over an element.
     *
     * @param element     the element to hover over
     * @param elementName a descriptive name for the element
     */
    @Step("Hover over: {elementName}")
    protected void hover(WebElement element, String elementName) {
        syncHelper.waitForElementVisible(element);
        log.debug("Hovering over: {}", elementName);
        new Actions(driver).moveToElement(element).perform();
    }

    /**
     * Drags an element and drops it on another element.
     *
     * @param source     the source element
     * @param target     the target element
     * @param sourceName the source element name
     * @param targetName the target element name
     */
    @Step("Drag {sourceName} to {targetName}")
    protected void dragAndDrop(WebElement source, WebElement target,
                               String sourceName, String targetName) {
        syncHelper.waitForElementVisible(source);
        syncHelper.waitForElementVisible(target);
        log.debug("Dragging {} to {}", sourceName, targetName);
        new Actions(driver).dragAndDrop(source, target).perform();
    }

    /**
     * Scrolls to an element.
     *
     * @param element     the element to scroll to
     * @param elementName a descriptive name for the element
     */
    @Step("Scroll to: {elementName}")
    protected void scrollToElement(WebElement element, String elementName) {
        log.debug("Scrolling to: {}", elementName);
        ((JavascriptExecutor) driver).executeScript(
                "arguments[0].scrollIntoView({behavior: 'smooth', block: 'center'});",
                element
        );
    }

    /**
     * Clicks an element using JavaScript.
     *
     * @param element     the element to click
     * @param elementName a descriptive name for the element
     */
    @Step("JS Click on: {elementName}")
    protected void jsClick(WebElement element, String elementName) {
        log.debug("JS clicking on: {}", elementName);
        ((JavascriptExecutor) driver).executeScript("arguments[0].click();", element);
    }

    /**
     * Waits for the page to fully load.
     */
    protected void waitForPageLoad() {
        syncHelper.waitForPageLoad();
    }

    /**
     * Takes a screenshot of the current page.
     *
     * @return the screenshot as bytes
     */
    protected byte[] takeScreenshot() {
        return DriverManager.takeScreenshot();
    }

    /**
     * Switches to a frame by index.
     *
     * @param index the frame index
     */
    @Step("Switch to frame {index}")
    protected void switchToFrame(int index) {
        log.debug("Switching to frame: {}", index);
        driver.switchTo().frame(index);
    }

    /**
     * Switches to a frame by name or ID.
     *
     * @param nameOrId the frame name or ID
     */
    @Step("Switch to frame: {nameOrId}")
    protected void switchToFrame(String nameOrId) {
        log.debug("Switching to frame: {}", nameOrId);
        driver.switchTo().frame(nameOrId);
    }

    /**
     * Switches to a frame by element.
     *
     * @param element the frame element
     */
    protected void switchToFrame(WebElement element) {
        log.debug("Switching to frame element");
        driver.switchTo().frame(element);
    }

    /**
     * Switches back to the default content.
     */
    @Step("Switch to default content")
    protected void switchToDefaultContent() {
        log.debug("Switching to default content");
        driver.switchTo().defaultContent();
    }

    /**
     * Switches to a new window/tab.
     *
     * @return the original window handle
     */
    protected String switchToNewWindow() {
        String originalWindow = driver.getWindowHandle();
        for (String windowHandle : driver.getWindowHandles()) {
            if (!windowHandle.equals(originalWindow)) {
                driver.switchTo().window(windowHandle);
                break;
            }
        }
        return originalWindow;
    }

    /**
     * Switches back to the original window.
     *
     * @param windowHandle the window handle to switch to
     */
    protected void switchToWindow(String windowHandle) {
        driver.switchTo().window(windowHandle);
    }

    /**
     * Accepts an alert.
     */
    @Step("Accept alert")
    protected void acceptAlert() {
        syncHelper.waitForAlert();
        driver.switchTo().alert().accept();
    }

    /**
     * Dismisses an alert.
     */
    @Step("Dismiss alert")
    protected void dismissAlert() {
        syncHelper.waitForAlert();
        driver.switchTo().alert().dismiss();
    }

    /**
     * Gets the text of an alert.
     *
     * @return the alert text
     */
    protected String getAlertText() {
        syncHelper.waitForAlert();
        return driver.switchTo().alert().getText();
    }

    /**
     * Sends keys to an alert prompt.
     *
     * @param text the text to send
     */
    @Step("Type '{text}' into alert")
    protected void sendKeysToAlert(String text) {
        syncHelper.waitForAlert();
        driver.switchTo().alert().sendKeys(text);
    }
}

File: js-module/playwright/src/pages/BasePage.ts
import { Page, Locator, expect } from '@playwright/test';
import { SelfHealingLocator } from '../ai/SelfHealingLocator';

/**
 * Base page object class providing common functionality for all page objects.
 * Includes self-healing locator support and comprehensive element interaction methods.
 */
export abstract class BasePage {
  readonly page: Page;
  readonly selfHealingLocator: SelfHealingLocator;

  constructor(page: Page) {
    this.page = page;
    this.selfHealingLocator = new SelfHealingLocator(page);
  }

  /**
   * Gets the page title.
   */
  async getPageTitle(): Promise<string> {
    return await this.page.title();
  }

  /**
   * Gets the current URL.
   */
  getCurrentUrl(): string {
    return this.page.url();
  }

  /**
   * Navigates to a URL.
   * @param url - The URL to navigate to
   */
  async navigateTo(url: string): Promise<void> {
    await this.page.goto(url);
    await this.waitForPageLoad();
  }

  /**
   * Navigates to a path relative to the base URL.
   * @param path - The relative path
   */
  async navigateToPath(path: string): Promise<void> {
    await this.page.goto(path);
    await this.waitForPageLoad();
  }

  /**
   * Refreshes the current page.
   */
  async refreshPage(): Promise<void> {
    await this.page.reload();
    await this.waitForPageLoad();
  }

  /**
   * Navigates back.
   */
  async goBack(): Promise<void> {
    await this.page.goBack();
  }

  /**
   * Navigates forward.
   */
  async goForward(): Promise<void> {
    await this.page.goForward();
  }

  /**
   * Waits for the page to fully load.
   */
  async waitForPageLoad(): Promise<void> {
    await this.page.waitForLoadState('domcontentloaded');
    await this.page.waitForLoadState('networkidle').catch(() => {
      // Network idle may timeout on some pages
    });
  }

  /**
   * Clicks on an element.
   * @param locator - The element locator
   */
  async click(locator: Locator): Promise<void> {
    await locator.click();
  }

  /**
   * Double-clicks on an element.
   * @param locator - The element locator
   */
  async doubleClick(locator: Locator): Promise<void> {
    await locator.dblclick();
  }

  /**
   * Right-clicks on an element.
   * @param locator - The element locator
   */
  async rightClick(locator: Locator): Promise<void> {
    await locator.click({ button: 'right' });
  }

  /**
   * Types text into an element.
   * @param locator - The element locator
   * @param text - The text to type
   * @param options - Optional typing options
   */
  async type(
    locator: Locator,
    text: string,
    options?: { clear?: boolean; delay?: number }
  ): Promise<void> {
    if (options?.clear !== false) {
      await locator.clear();
    }
    await locator.fill(text);
  }

  /**
   * Types text character by character (simulates real typing).
   * @param locator - The element locator
   * @param text - The text to type
   * @param delay - Delay between keystrokes in ms
   */
  async typeSlowly(locator: Locator, text: string, delay = 50): Promise<void> {
    await locator.clear();
    await locator.pressSequentially(text, { delay });
  }

  /**
   * Gets the text content of an element.
   * @param locator - The element locator
   */
  async getText(locator: Locator): Promise<string> {
    return (await locator.textContent()) || '';
  }

  /**
   * Gets the inner text of an element.
   * @param locator - The element locator
   */
  async getInnerText(locator: Locator): Promise<string> {
    return await locator.innerText();
  }

  /**
   * Gets the value of an input element.
   * @param locator - The element locator
   */
  async getValue(locator: Locator): Promise<string> {
    return await locator.inputValue();
  }

  /**
   * Gets an attribute value of an element.
   * @param locator - The element locator
   * @param attribute - The attribute name
   */
  async getAttribute(locator: Locator, attribute: string): Promise<string | null> {
    return await locator.getAttribute(attribute);
  }

  /**
   * Checks if an element is visible.
   * @param locator - The element locator
   */
  async isVisible(locator: Locator): Promise<boolean> {
    return await locator.isVisible();
  }

  /**
   * Checks if an element is enabled.
   * @param locator - The element locator
   */
  async isEnabled(locator: Locator): Promise<boolean> {
    return await locator.isEnabled();
  }

  /**
   * Checks if a checkbox/radio is checked.
   * @param locator - The element locator
   */
  async isChecked(locator: Locator): Promise<boolean> {
    return await locator.isChecked();
  }

  /**
   * Selects an option from a dropdown by value.
   * @param locator - The select element locator
   * @param value - The option value
   */
  async selectByValue(locator: Locator, value: string): Promise<void> {
    await locator.selectOption({ value });
  }

  /**
   * Selects an option from a dropdown by label.
   * @param locator - The select element locator
   * @param label - The option label
   */
  async selectByLabel(locator: Locator, label: string): Promise<void> {
    await locator.selectOption({ label });
  }

  /**
   * Selects an option from a dropdown by index.
   * @param locator - The select element locator
   * @param index - The option index
   */
  async selectByIndex(locator: Locator, index: number): Promise<void> {
    await locator.selectOption({ index });
  }

  /**
   * Checks a checkbox.
   * @param locator - The checkbox locator
   */
  async check(locator: Locator): Promise<void> {
    await locator.check();
  }

  /**
   * Unchecks a checkbox.
   * @param locator - The checkbox locator
   */
  async uncheck(locator: Locator): Promise<void> {
    await locator.uncheck();
  }

  /**
   * Hovers over an element.
   * @param locator - The element locator
   */
  async hover(locator: Locator): Promise<void> {
    await locator.hover();
  }

  /**
   * Drags an element and drops it on another.
   * @param source - The source element locator
   * @param target - The target element locator
   */
  async dragAndDrop(source: Locator, target: Locator): Promise<void> {
    await source.dragTo(target);
  }

  /**
   * Scrolls an element into view.
   * @param locator - The element locator
   */
  async scrollIntoView(locator: Locator): Promise<void> {
    await locator.scrollIntoViewIfNeeded();
  }

  /**
   * Waits for an element to be visible.
   * @param locator - The element locator
   * @param timeout - Optional timeout in ms
   */
  async waitForVisible(locator: Locator, timeout?: number): Promise<void> {
    await locator.waitFor({ state: 'visible', timeout });
  }

  /**
   * Waits for an element to be hidden.
   * @param locator - The element locator
   * @param timeout - Optional timeout in ms
   */
  async waitForHidden(locator: Locator, timeout?: number): Promise<void> {
    await locator.waitFor({ state: 'hidden', timeout });
  }

  /**
   * Takes a screenshot of the page.
   * @param name - Optional screenshot name
   */
  async takeScreenshot(name?: string): Promise<Buffer> {
    return await this.page.screenshot({
      fullPage: true,
      path: name ? `screenshots/${name}.png` : undefined,
    });
  }

  /**
   * Takes a screenshot of a specific element.
   * @param locator - The element locator
   * @param name - Optional screenshot name
   */
  async takeElementScreenshot(locator: Locator, name?: string): Promise<Buffer> {
    return await locator.screenshot({
      path: name ? `screenshots/${name}.png` : undefined,
    });
  }

  /**
   * Presses a key or key combination.
   * @param key - The key to press (e.g., 'Enter', 'Control+A')
   */
  async pressKey(key: string): Promise<void> {
    await this.page.keyboard.press(key);
  }

  /**
   * Waits for a specific URL pattern.
   * @param urlPattern - The URL pattern to wait for
   * @param timeout - Optional timeout in ms
   */
  async waitForUrl(urlPattern: string | RegExp, timeout?: number): Promise<void> {
    await this.page.waitForURL(urlPattern, { timeout });
  }

  /**
   * Waits for a network request to complete.
   * @param urlPattern - The URL pattern to match
   */
  async waitForRequest(urlPattern: string | RegExp): Promise<void> {
    await this.page.waitForRequest(urlPattern);
  }

  /**
   * Waits for a network response.
   * @param urlPattern - The URL pattern to match
   */
  async waitForResponse(urlPattern: string | RegExp): Promise<void> {
    await this.page.waitForResponse(urlPattern);
  }

  /**
   * Evaluates JavaScript in the page context.
   * @param pageFunction - The function to evaluate
   */
  async evaluate<T>(pageFunction: () => T): Promise<T> {
    return await this.page.evaluate(pageFunction);
  }

  /**
   * Gets the count of elements matching a locator.
   * @param locator - The element locator
   */
  async getElementCount(locator: Locator): Promise<number> {
    return await locator.count();
  }

  /**
   * Gets all texts from elements matching a locator.
   * @param locator - The element locator
   */
  async getAllTexts(locator: Locator): Promise<string[]> {
    return await locator.allTextContents();
  }

  /**
   * Uploads a file to an input element.
   * @param locator - The file input locator
   * @param filePath - The path to the file
   */
  async uploadFile(locator: Locator, filePath: string | string[]): Promise<void> {
    await locator.setInputFiles(filePath);
  }

  /**
   * Handles a dialog (alert, confirm, prompt).
   * @param action - The action to take ('accept' or 'dismiss')
   * @param promptText - Optional text to enter for prompts
   */
  async handleDialog(action: 'accept' | 'dismiss', promptText?: string): Promise<void> {
    this.page.once('dialog', async (dialog) => {
      if (action === 'accept') {
        await dialog.accept(promptText);
      } else {
        await dialog.dismiss();
      }
    });
  }

  /**
   * Asserts that an element is visible.
   * @param locator - The element locator
   */
  async assertVisible(locator: Locator): Promise<void> {
    await expect(locator).toBeVisible();
  }

  /**
   * Asserts that an element contains specific text.
   * @param locator - The element locator
   * @param text - The expected text
   */
  async assertContainsText(locator: Locator, text: string): Promise<void> {
    await expect(locator).toContainText(text);
  }

  /**
   * Asserts that the page title matches.
   * @param title - The expected title
   */
  async assertTitle(title: string | RegExp): Promise<void> {
    await expect(this.page).toHaveTitle(title);
  }

  /**
   * Asserts that the URL matches.
   * @param url - The expected URL
   */
  async assertUrl(url: string | RegExp): Promise<void> {
    await expect(this.page).toHaveURL(url);
  }
}
